core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/LogicalPlan.scala

```diff
@@ -143,119 +128,81 @@ case class LogicalPlan(
       val op = getOperator(opId)
-      // infer output schema of this operator based on its input schema
-      val outputSchemas: Option[Array[Schema]] = {
-        // call to "getOutputSchema" might cause exceptions, wrap in try/catch and return empty schema
-        try {
-          if (op.isInstanceOf[SourceOperatorDescriptor]) {
-            // op is a source operator, ask for it output schema
-            Option.apply(op.getOutputSchemas(Array()))
-          } else if (
-            !inputSchemaMap.contains(op.operatorIdentifier) || inputSchemaMap(op.operatorIdentifier)
-              .exists(s => s.isEmpty)
-          ) {
-            // op does not have input, or any of the op's input's output schema is null
-            // then this op's output schema cannot be inferred as well
-            Option.empty
-          } else {
-            // op's input schema is complete, try to infer its output schema
-            // if inference failed, print an exception message, but still continue the process
-            Option.apply(
-              op.getOutputSchemas(inputSchemaMap(op.operatorIdentifier).map(s => s.get).toArray)
+      val inputSchemas: Array[Option[Schema]] = if (op.isInstanceOf[SourceOperatorDescriptor]) {
+        Array()
+      } else {
+        op.operatorInfo.inputPorts
+          .flatMap(inputPort => {
+            links
+              .filter(link => link.toOpId == op.operatorIdentifier && link.toPortId == inputPort.id)
+              .map(link => {
+                val outputSchemaOpt =
+                  getOperator(link.fromOpId).outputPortToSchemaMapping.get(link.fromPortId)
+                if (outputSchemaOpt.isDefined) {
+                  op.inputPortToSchemaMapping(inputPort.id) = outputSchemaOpt.get
+                }
+                outputSchemaOpt
+              })
+          })
+          .toArray
+      }
+
+      if (!inputSchemas.contains(None)) {
+        Try(op.getOutputSchemas(inputSchemas.map(_.get))) match {
+          case Success(outputSchemas) =>
+            op.operatorInfo.outputPorts.foreach(outputPort =>
+              op.outputPortToSchemaMapping(outputPort.id) = outputSchemas(outputPort.id.id)
             )
-          }
-        } catch {
-          case e: Throwable =>
-            logger.error("got error", e)
+
+            if (!op.isInstanceOf[SinkOpDesc] && outputSchemas.nonEmpty) {
+              assert(outputSchemas.length == op.operatorInfo.outputPorts.length)
+            }
+
+          case Failure(err) =>
+            logger.error("got error", err)
             errorList match {
-              case Some(list) => list.append((opId, e))
+              case Some(list) => list.append((opId, err))
               case None       =>
             }
-
-            Option.empty
         }
-      }
 
-      // exception: if op is a source operator, use its output schema as input schema for autocomplete
-      if (op.isInstanceOf[SourceOperatorDescriptor]) {
```

Comment:
just curious, is this logic merged into `Try(op.getOutputSchemas(inputSchemas.map(_.get)))`?